File: tft_webportal/templates/includes/transportation_request.js
const placeState = {
	"from": {"address": {}, "phone": "", "company": "", "from_time": "", "to_time": ""},
	"to": {"address": {}, "phone": "", "company": "", "from_time": "", "to_time": ""}
}


frappe.ready(() => {
	const today = frappe.datetime.get_today();
	$("#request_date").val(today);
	$("#pickup_date").val(today);
	$("#delivery_date").val(today);

	/* view request */
	view_request();
	const form = document.querySelector(".tft-transport form");
	if (!form) {return;}

	/* Vehicle rows dynamic add/remove */
	const rows = document.getElementById("vehicle-rows");
	const addBtn = document.querySelector("[data-add-vehicle-row]");
	if (!rows || !addBtn) {
		return;
	}

	let rowIndex = rows.children.length;

	function buildCell(name, type = "text", placeholder = "", datalist = null) {
		const td = document.createElement("td");
		const input = document.createElement("input");
		input.name = name;
		input.type = type;
		if (placeholder) {
			input.placeholder = placeholder;
		}
		if (datalist) {
			input.setAttribute("list", datalist);
			if(datalist === "vin_list"){
				input.setAttribute("onchange", "update_vehicle_info(this)");
			}
		}
		if (type === "number") {
			input.min = "1";
			input.placeholder = placeholder || "1";
		}
		td.appendChild(input);
		return td;
	}

	function buildSelectCell(name, options, defaultValue = null) {
		const td = document.createElement("td");
		const select = document.createElement("select");

		select.name = name;

		options.forEach(opt => {
			const option = document.createElement("option");
			option.value = opt.value;
			option.textContent = opt.label;

			if (defaultValue !== null && opt.value === defaultValue) {
				option.selected = true;
			}

			select.appendChild(option);
		});

		td.appendChild(select);
		return td;
	}

	function buildRemoveCell() {
		const td = document.createElement("td");
		td.className = "row-actions";
		const btn = document.createElement("button");
		btn.type = "button";
		btn.className = "row-remove";
		btn.setAttribute("aria-label", "Remove row");
		btn.textContent = "✕";
		td.appendChild(btn);
		return td;
	}

	addBtn.addEventListener("click", () => {
		rowIndex += 1;
		const tr = document.createElement("tr");
		tr.appendChild(buildCell(`vin_${rowIndex}`, "text", "1C3CCBAB5EN125414", "vin_list"));
		tr.appendChild(buildCell(`vehicle_number_${rowIndex}`, "text", "4521909"));
		tr.appendChild(buildCell(`make_model_${rowIndex}`, "text", "2014 Chrysler 200"));
		tr.appendChild(buildCell(`description_${rowIndex}`, "text", "Car", "description_list"));
		tr.appendChild(buildCell(`plate_${rowIndex}`, "text", "CTET844"));
		tr.appendChild(buildSelectCell(`hitch_${rowIndex}`, [{ label: "No", value: "No" }, { label: "Yes", value: "Yes" }], "No"));
		tr.appendChild(buildRemoveCell());
		rows.appendChild(tr);
	});

	rows.addEventListener("click", (event) => {
		const btn = event.target.closest(".row-remove");
		if (!btn) {
			return;
		}
		const row = btn.closest("tr");
		if (!row) {
			return;
		}
		if (rows.children.length <= 1) {
			row.querySelectorAll("input").forEach((input) => {
				input.value = "";
			});
			return;
		}
		row.remove();
	});

	const addressInputs = document.querySelectorAll(".address-input[list]");
	addressInputs.forEach((input) => {
		const listId = input.getAttribute("list");
		if (!listId) {
			return;
		}
		input.dataset.listId = listId;
		input.addEventListener("input", () => {
			if (input.value.trim()) {
				input.removeAttribute("list");
			} else {
				input.setAttribute("list", listId);
			}
		});
		input.addEventListener("blur", () => {
			input.setAttribute("list", listId);
		});
	});
	/* End vehicle rows dynamic add/remove */
});

/* view request */
function view_request() {
	// Load existing request data if in view mode
	const urlParams = new URLSearchParams(window.location.search);
	const requestName = urlParams.get("request");
	const viewMode = urlParams.get("view") === "1";
	if (!requestName) return;
	
	// Load request data
	frappe.call({
		method: "tft_webportal.www.transportation_request.get_request",
		args: { request: requestName },
		callback: function(r) {
			const request = r.message;
			if (!request) return;

			// Fill fields
			$("#supplier_email").val(request.supplier);
			$("#contact_email").val(request.supplier);
			$("#request_date").val(request.date);
			$("#po_number").val(request.po_number);
			$("#business_unit").val(request.business_unit_name);
			$("#contact_name").val(request.contact_name);
			$("#contact_phone").val(request.contact_phone);
			$("#pickup_date").val(request.available_pickup_date);
			$("#delivery_date").val(request.delivery_required_by);
			$("#notes").val(request.notes);

			// Fill addresses
			$("#from_company").val(request.addresses[0].from_company);
			$("#from_address").val(request.addresses[0].from_address);
			$("#from_contact").val(request.addresses[0].from_contact);
			$("#from_start_hour").val(request.addresses[0].from_start_time);
			$("#from_end_hour").val(request.addresses[0].from_end_time);
			$("#from_gate").val(request.addresses[0].from_code);

			$("#to_company").val(request.addresses[0].to_company);
			$("#to_address").val(request.addresses[0].to_address);
			$("#to_contact").val(request.addresses[0].to_contact);
			$("#to_start_hour").val(request.addresses[0].to_start_time);
			$("#to_end_hour").val(request.addresses[0].to_end_time);
			$("#to_gate").val(request.addresses[0].to_code);

			// Fill vehicles dynamically
			const $vehicleRows = $("#vehicle-rows");
			$vehicleRows.empty();
			request.vehicle.forEach((v, index) => {
				const rowIndex = index + 1;
				const hitchValue = v.cables ? "Yes" : "No";
				const row = `<tr>
					<td><input type="text" name="vin_${rowIndex}" value="${v.vin}"></td>
					<td><input type="text" name="vehicle_number_${rowIndex}" value="${v.vehicle_number}"></td>
					<td><input type="text" name="make_model_${rowIndex}" value="${v.make}"></td>
					<td><input type="text" name="description_${rowIndex}" value="${v.type}"></td>
					<td><input type="text" name="plate_${rowIndex}" value="${v.plate}"></td>
					<td>
						<select name="hitch_${rowIndex}">
							<option value="No" ${hitchValue === "No" ? "selected" : ""}>No</option>
							<option value="Yes" ${hitchValue === "Yes" ? "selected" : ""}>Yes</option>
						</select>
					</td>
					<td class="row-actions"><button type="button" class="row-remove" aria-label="Remove row">✕</button></td>
				</tr>`;
				$vehicleRows.append(row);
			});

			// Hide submit button in view mode
			if (viewMode) {
				const addBtn = document.querySelector("[data-add-vehicle-row]");
				$("#submit_request").hide();
				$(addBtn).hide();
				$(".row-remove").hide();
			}
		}
	});
}

/* End view request */

/* address autocomplete */
const address_list = {{ addresses | safe }};
function update_contact_info(input_element) {
	const value = input_element.value;
	for (let i in address_list) {
		if (address_list[i].name === value) {
			document.getElementById(input_element.name === "from_address" ? "from_contact" : "to_contact").value = address_list[i].phone || '';
			document.getElementById(input_element.name === "from_address" ? "from_company" : "to_company").value = address_list[i].company || '';
			document.getElementById(input_element.name === "from_address" ? "from_start_hour" : "to_start_hour").value = address_list[i].from_time || '08:00';
			document.getElementById(input_element.name === "from_address" ? "from_end_hour" : "to_end_hour").value = address_list[i].to_time || '17:00';
		}
	}
}

function initAutocomplete() {
	/* from address autocomplete */
	const from_address_input = document.getElementById("from_address");
	const from_address_autocomplete = new google.maps.places.Autocomplete(from_address_input, {componentRestrictions: {country: ["us", "ca"]}, types: ['address']});

	from_address_autocomplete.addListener("place_changed", () => {
		const place = from_address_autocomplete.getPlace();
		document.getElementById("from_address").value = place.formatted_address;
		placeState.from.address = mapGooglePlace(place);
	});

	/* to address autocomplete */
	const to_address_input = document.getElementById("to_address");
	const to_address_autocomplete = new google.maps.places.Autocomplete(to_address_input, {componentRestrictions: {country: ["us", "ca"]}, types: ['address']});
	to_address_autocomplete.addListener("place_changed", () => {
		const place = to_address_autocomplete.getPlace();
		document.getElementById("to_address").value = place.formatted_address;
		placeState.to.address = mapGooglePlace(place);
	});
}

function mapGooglePlace(place) {
	const get = (type, short = false) =>
		place.address_components.find(c => c.types.includes(type))?.[short ? "short_name" : "long_name"] || "";

	return {
		address_line1: [get("street_number"), get("route")].filter(Boolean).join(" "),
		city: get("locality") || get("sublocality"),
		state: get("administrative_area_level_1", true),
		pincode: get("postal_code"),
		country: get("country")
	};
}

/* end address autocomplete */

/* vehicle data resource */
const vehicle_list = {{ vehicles | safe }};
function update_vehicle_info(input_element) {
	const value = input_element.value;
	for (let i in vehicle_list) {
		if (vehicle_list[i].vin === value) {
			const rowIndex = input_element.name.split('_')[1];
			document.getElementsByName(`vehicle_number_${rowIndex}`)[0].value = vehicle_list[i].vehicle_number || '';
			document.getElementsByName(`make_model_${rowIndex}`)[0].value = vehicle_list[i].make_model || '';
			document.getElementsByName(`description_${rowIndex}`)[0].value = vehicle_list[i].description || '';
			document.getElementsByName(`plate_${rowIndex}`)[0].value = vehicle_list[i].plate || '';
			document.getElementsByName(`hitch_${rowIndex}`)[0].value = vehicle_list[i].hitch ? "Yes" : "No";
		}
	}
}

/* end vehicle data resource */

/* submit request */
$("#request-form").on("submit", function (e) {
	e.preventDefault();

	/* validations */
	if(!$("#from_address").val() || !$("#to_address").val()){
		frappe.msgprint(__("Please select both From and To addresses."));
		return;
	}

	if($("#from_address").val() === $("#to_address").val()){
		frappe.msgprint(__("From and To addresses cannot be the same."));
		return;
	}

	const vins = new Set();
	const plates = new Set();

	for (const row of document.querySelectorAll("#vehicle-rows tr")) {
		const vin = row.querySelector("input[name^='vin_']").value.trim().toUpperCase();
		const plate = row.querySelector("input[name^='plate_']").value.trim().toUpperCase();
		
		if (vin) {
			if (vins.has(vin)) {
				frappe.msgprint(__("Duplicate VIN found: {0}".format(vin)));
				return;
			}
			vins.add(vin);
		}
		
		if (plate) {
			if (plates.has(plate)) {
				frappe.msgprint(__("Duplicate Plate found: {0}".format(plate)));
				return;
			}
			plates.add(plate);
		}
	}

	/* place data */
	placeState.from.phone = $("[name='from_contact']").val();
	placeState.from.company = $("[name='from_company']").val();
	placeState.from.from_time = $("[name='from_start_hour']").val();
	placeState.from.to_time = $("[name='from_end_hour']").val();

	placeState.to.phone = $("[name='to_contact']").val();
	placeState.to.company = $("[name='to_company']").val();
	placeState.to.from_time = $("[name='to_start_hour']").val();
	placeState.to.to_time = $("[name='to_end_hour']").val();

	/* vehicle data */
	const vehicles = [];
	$("#vehicle-rows tr").each(function (index) {
		const row_index = index + 1;
		const hitch_val = $(this).find(`select[name="hitch_${row_index}"]`).val();
		const hitch_boolean = hitch_val === "Yes" ? 1 : 0;

		vehicles.push({
			vehicle_number: $(this).find(`input[name="vehicle_number_${row_index}"]`).val(),
			vin: $(this).find(`input[name="vin_${row_index}"]`).val().toUpperCase(),
			make_model: $(this).find(`input[name="make_model_${row_index}"]`).val(),
			description: $(this).find(`input[name="description_${row_index}"]`).val(),
			plate: $(this).find(`input[name="plate_${row_index}"]`).val().toUpperCase(),
			hitch: hitch_boolean
		});
	});

	const data = {
		request_date: $("[name='request_date']").val(), po_number: $("[name='po_number']").val(),
		business_unit: $("[name='business_unit']").val(), contact_name: $("[name='contact_name']").val(),
		contact_phone: $("[name='contact_phone']").val(), pickup_date: $("[name='pickup_date']").val(),
		delivery_date: $("[name='delivery_date']").val(), notes: $("[name='notes']").val(),
		from_company: $("[name='from_company']").val(), from_address: $("[name='from_address']").val(),
		from_contact: $("[name='from_contact']").val(), from_start_hour: $("[name='from_start_hour']").val(),
		from_end_hour: $("[name='from_end_hour']").val(), from_gate: $("[name='from_gate']").val(),
		to_company: $("[name='to_company']").val(), to_address: $("[name='to_address']").val(),
		to_contact: $("[name='to_contact']").val(), to_start_hour: $("[name='to_start_hour']").val(),
		to_end_hour: $("[name='to_end_hour']").val(), to_gate: $("[name='to_gate']").val(),
		vehicles: vehicles,	addresses: placeState
	};

	warning = ""
	if (!data.po_number && !data.notes){
		warning = __("Fields <b>PO Number</b> and <b>Notes</b> for instruction are left empty");
	} else if (!data.po_number){
		warning = __("Field <b>PO Number</b> is left Empty");
	}else if (!data.notes){
		warning = __("Field <b>Notes</b> for instruction is left Empty");
	}

	if(warning != ""){
		frappe.warn(__('Are you sure you want to proceed?'), warning,
			() => {
				create_request(data)
			},
			'Continue',
			true
		)
	} else {
		create_request(data);
	}
});

function create_request(data){
	frappe.call({
		method: "tft_webportal.www.transportation_request.create_request",
		args: {data: data},
		freeze: true,
		freeze_message: __("Creating Transportation Request..."),
		callback: function (r) {
			if (r.message) {
				frappe.show_alert(__(r.message.message), 300);
				setTimeout(() => {
					window.location.reload();
				}, 1000);
			}
		}
	});
}

/* end submit request */

File: tft_webportal/www/transportation_request.py
import frappe
from frappe import _
from frappe.utils import getdate

from datetime import datetime

def get_context(context):
	context.addresses = []
	context.vehicles = []
	if frappe.session.user == "Guest":
		frappe.throw(_("You need to be logged in to access this page"), frappe.PermissionError)

	user = frappe.get_doc("User", frappe.session.user, ignore_permissions=1)
	context.user_email = user.email
	context.user_phone = user.phone
	context.business_unit = user.company
	context.contact_name = user.full_name
	context.contact_phone = user.phone
	context.contact_email = user.email

	# Load addresses
	if user.approved_company:
		users = frappe.db.get_all("User", filters={"approved_company": user.approved_company}, pluck="name")
	else:
		users = [frappe.session.user]

	addresses = frappe.db.sql("""
		SELECT
			a.name, a.address_title, a.custom_company as company, a.custom_from_time as from_time, a.custom_to_time as to_time, 
			a.address_line1, a.city, a.state, a.pincode, a.country, a.phone, a.custom_formatted_address as formatted_address
		FROM `tabAddress` a
			INNER JOIN `tabDynamic Link` dl ON dl.parent = a.name
			WHERE dl.link_name IN %(users)s
			AND dl.link_doctype = "User"
		""", {"users": users}, as_dict=1)

	for addr in addresses:
		if addr.company == None: addr.company = ""
		
		addr.from_time = datetime.strptime(str(addr.from_time), "%H:%M:%S").strftime("%H:%M") if addr.from_time else ""
		addr.to_time = datetime.strptime(str(addr.to_time), "%H:%M:%S").strftime("%H:%M") if addr.to_time else ""
		addr.display = ", ".join(filter(None, [addr.company, addr.address_line1, addr.city, addr.state, addr.pincode, addr.country]))
		context.addresses.append({
			"name": addr.formatted_address if addr.formatted_address else addr.address_line1, 
			"display": addr.display, "phone": addr.phone, "company": addr.company, "from_time": addr.from_time, "to_time": addr.to_time
		})


	# Load vehicles
	vehicles = frappe.db.get_all("Client Vehicle", filters={"user": ["in", users]}, fields=["vehicle_number", "make", "type", "plate", "vin", "cables", "fuel_type"])
	for vehicle in vehicles:
		vehicle.display = ", ".join(filter(None, [vehicle.vehicle_number, vehicle.make, vehicle.type, vehicle.plate]))
		context.vehicles.append({
			"vin": vehicle.vin, "display": vehicle.display, "vehicle_number": vehicle.vehicle_number, "make_model": vehicle.make, 
			"description": vehicle.type, "plate": vehicle.plate, "hitch": vehicle.cables, "fuel_type": vehicle.fuel_type or ""
		})
	context.vehicle_description = frappe.get_all("Vehicle Type", pluck="name") or []

	# load fuel types
	context.fuel_types = frappe.get_all("Vehicle Fuel Type", pluck="name", ignore_permissions=1)

# ------------------------------------------------------------- #

# -------------- create transaction -------------- #
@frappe.whitelist()
def create_request(data):
	try:
		# Enforce login
		if frappe.session.user == "Guest": frappe.throw(_("Please login to continue"), frappe.PermissionError)

		supplier = frappe.session.user

		# Validate required fields
		payload = frappe.parse_json(data) if data else {}
		required_fields = ["pickup_date", "delivery_date", "from_address", "to_address"]
		missing = [field for field in required_fields if not payload.get(field)]

		# Address Resolution
		if missing: frappe.throw(_("Missing required fields: {0}.").format(", ".join(missing)))

		# Date validation
		pickup_dt = getdate(payload.get("pickup_date"))
		delivery_dt = getdate(payload.get("delivery_date"))
		
		# Date validation
		if delivery_dt < pickup_dt: frappe.throw(_("Delivery Required By Date cannot be earlier than Pickup Date."))

		# Vehicle Resolution
		if payload.get("from_address") == payload.get("to_address"):
			frappe.throw(_("From and To addresses cannot be the same."))

		# Address resolution
		addresses = payload.get("addresses")
		resolve_address(addresses.get("from"), supplier, payload.get("business_unit"))
		resolve_address(addresses.get("to"), supplier, payload.get("business_unit"))

		# Vehicle Resolution
		resolve_vehicle(payload, supplier)

		# Create Transportation Request document
		for vehicle in payload.get("vehicles"):
			job = frappe.get_doc({
				"doctype": "Transportation Request", "supplier": supplier,
				"p_o_number": vehicle.get("po_number"),	"date": payload.get("request_date"),
				"business_unit_name": payload.get("business_unit"),	"contact_name": payload.get("contact_name"),
				"contact_phone": payload.get("contact_phone"), "available_pickup_date": pickup_dt,
				"delivery_required_by": delivery_dt, "notes": payload.get("notes"),
				"from_company": payload.get("from_company"), "from_address": payload.get("from_address"),
				"from_contact": payload.get("from_contact"), "from_start_time": payload.get("from_start_hour"),
				"from_end_time": payload.get("from_end_hour"), "from_code": payload.get("from_gate"),
				"to_company": payload.get("to_company"), "to_address": payload.get("to_address"),
				"to_contact": payload.get("to_contact"), "to_start_time": payload.get("to_start_hour"),
				"to_end_time": payload.get("to_end_hour"), "to_code": payload.get("to_gate"),
				"vin": vehicle.get("vin"), "vehicle_number": vehicle.get("vehicle_number"),
				"make": vehicle.get("make_model"), "type": vehicle.get("description"),
				"fuel_type": vehicle.get("fuel_type"), "plate": vehicle.get("plate"), "cables": vehicle.get("hitch")
			})
			job.insert(ignore_permissions=True)

		return frappe._dict({"job": job.name, "message": _("Request(s) created successfully")})
	except Exception as e:
		frappe.error_log("Error", f'{e}')

#---------------- address resolution ----------------- #
def resolve_address(payload, username, business_unit=None):
	addr = payload.get("address")
	if not addr: return None
	
	# Check if address already exists
	existing = frappe.db.get_value("Address", {"address_line1": addr.get("address_line1"), "city": addr.get("city"), "pincode": addr.get("pincode")}, "name")
	if existing:
		users = frappe.db.get_all("User", filters={"approved_company": frappe.get_value("User", username, "approved_company")}, pluck="name")

		# Check if the address is already linked to the users
		link_exists = frappe.db.exists({"doctype": "Dynamic Link", "parent": existing, "link_doctype": "User", "link_name": ["in", users]})

		if not link_exists:
			link = {"link_doctype": "User", "link_name": username}
			address_doc = frappe.get_doc("Address", existing, ignore_permissions=True)
			address_doc.append("links", link)
			address_doc.save(ignore_permissions=True)
		return existing
	
	# Create a new address
	link = frappe._dict({"link_doctype": "User", "link_name": username})
	address = frappe.get_doc({
		"doctype": "Address", "address_type": "Shipping", "custom_formatted_address": addr.get("formatted_address"),
		"address_title": addr.get("address_line1"), "custom_company": payload.get("company") or business_unit,
		"phone": payload.get("phone"), "custom_from_time": payload.get("from_time"), "custom_to_time": payload.get("to_time"), **addr
	})
	address.append("links", link)
	address.insert(ignore_permissions=True)
	return address.name

#---------------- vehicle resolution ----------------- #
def resolve_vehicle(payload, username):
	# Validate duplicate VINs and Plates in the payload
	vin_set = set()
	plate_set = set()
	
	for row in payload.get("vehicles"):
		vin = row.get("vin")
		plate = row.get("plate")

		if vin in vin_set: frappe.throw(_("Duplicate VIN found: {0}").format(vin))
		vin_set.add(vin)
		
		if plate in plate_set: frappe.throw(_("Duplicate Plate found: {0}").format(plate))
		plate_set.add(plate)
		
	vehicles = payload.get("vehicles")
	users = frappe.db.get_all("User", filters={"approved_company": frappe.get_value("User", username, "approved_company")}, pluck="name")
	# Insert vehicles if they do not already exist
	for vehicle in vehicles:
		existing = frappe.db.get_value("Client Vehicle", {"vin": vehicle.get("vin"), "user": ["in", users]}, ["name", "plate"])
		if existing and existing[1] == vehicle.get("plate"):
			continue
		elif existing and existing[1] != vehicle.get("plate"):
			frappe.throw(_("VIN {0} is already associated with a different plate.").format(vehicle.get('vin')))
			return
		
		veh = frappe.get_doc({
			"make": vehicle.get("make_model"), "type": vehicle.get("description"),
			"doctype": "Client Vehicle", "user": username, "vehicle_number": vehicle.get("vehicle_number"),
			"plate": vehicle.get("plate"), "vin": vehicle.get("vin"), "cables": vehicle.get("hitch"), "fuel_type": vehicle.get("fuel_type"),
		})
		veh.insert(ignore_permissions=True)

# ---------------- end ----------------- #

# ------------------ get request ------------------ #
@frappe.whitelist()
def get_request(request_id):
	# Enforce login
	if frappe.session.user == "Guest": frappe.throw(_("Please login to continue"), frappe.PermissionError)
	request = frappe.get_doc("Transportation Request", request_id, ignore_permissions=True)
	request.from_start_time = datetime.strptime(str(request.from_start_time), "%H:%M:%S").strftime("%H:%M") if request.from_start_time else ""
	request.from_end_time = datetime.strptime(str(request.from_end_time), "%H:%M:%S").strftime("%H:%M") if request.from_end_time else ""
	request.to_start_time = datetime.strptime(str(request.to_start_time), "%H:%M:%S").strftime("%H:%M") if request.to_start_time else ""
	request.to_end_time = datetime.strptime(str(request.to_end_time), "%H:%M:%S").strftime("%H:%M") if request.to_end_time else ""
	return request.as_dict()